src/scores/continuous/nse_impl.py

```diff
@@ -0,0 +1,494 @@
+        obs=metads.datasets.obs,
+        weights=metads.datasets.weights,
+        reduce_dims=reduce_dims,
+        preserve_dims=preserve_dims,
+        is_angular=is_angular,
+    )
+
+    # executes the NseScore object and resolves the appropriate return datatype
+    nse_result: XarrayLike = NseUtils.run_nsescore_and_undo_metadataset(
+        obj_nsescore,
+        metads,
+    )
+
+    # safety: assert xarraylike before returning
+    assert is_xarraylike(nse_result)
+
+    return nse_result
+
+
+@dataclass(kw_only=True)
+class NseScore:
+    """
+    Helper structure used to perform checks on data during initialization.
+    """
+
+    #: forecast/prediction/simulation/model
+    fcst: xr.Dataset
+    #: observations
+    obs: xr.Dataset
+    #: weighting
+    weights: xr.Dataset | None = None
+    #: dimensions to reduce
+    reduce_dims: FlexibleDimensionTypes | None = None
+    #: dimensions to preserve (mutually exlcusive to reduce_dims)
+    preserve_dims: FlexibleDimensionTypes | None = None
+    #: whether the input data is angular, used implicitly by `mse`
+    is_angular: bool = False
+
+    def __post_init__(self):
+        """
+        Run checks before creating final object
+            - runtime type assertions
+            - check weights
+            - dimension checks
+        """
+        # safety: dev/testing: runtime assertions on arguments
+        assert (
+            isinstance(self.fcst, xr.Dataset)
+            and isinstance(self.obs, xr.Dataset)
+            # These are optionals (i.e. skipped if None):
+            and (self.weights is None or isinstance(self.weights, xr.Dataset))
+            and (self.reduce_dims is None or is_flexibledimensiontypes(self.reduce_dims))
+            and (self.preserve_dims is None or is_flexibledimensiontypes(self.preserve_dims))
+            and isinstance(self.is_angular, bool)
+        )
+
+        # check weights if its provided
+        if self.weights is not None:
+            check_weights(self.weights)
+
+        # do any dimension checks and store the final reduced (gathered) dimensions
+        self.reduce_dims = NseUtils.check_and_gather_dimensions(
+            fcst=self.fcst,
+            obs=self.obs,
+            weights=self.weights,
+            reduce_dims=self.reduce_dims,
+            preserve_dims=self.preserve_dims,
+        )
+
+    # The following methods are defined as "properties", since a NseScore object should
+    # not be mutable so any computation should also only act as a "getter"
+
+    @property
+    def nse(self) -> xr.Dataset:
+        """
+        Actual ``nse`` computation happens here.
+
+        Divide by zero error is ignored as it will be already raised by call to
+        ``obs_variance`` computation.
+        """
+        with np.errstate(divide="ignore"):
+            ret_nse: xr.Dataset = 1.0 - (self.fcst_error / self.obs_variance)
+            return ret_nse
+
+    @property
+    def fcst_error(self) -> XarrayLike:
```

Comment:
Can use more specific type hinting: `NseScore` should only work with datasets

```suggestion
    def fcst_error(self) -> xr.Dataset:
```